docs/source/plugins/for_plugin_developers.rst

```diff
@@ -0,0 +1,274 @@
+
+2. **Hook Specifications**:  We then create "*hook specifications*", which are
+little more than well-documented function signatures that define the API (or
+"contract") that a plugin developer must adhere to when writing their function
+that we promise to call somewhere in the napari codebase.
+
+3. **Hook Implementations**: Plugin developers then write functions ("*hook
+implementations*") and mark that function as meeting the requirements of a
+specific *hook specification* offered by napari (using a decorator as
+:ref:`described below <hookimpl-decorator>`).
+
+4. **Plugin discovery**: Using :ref:`one of two methods <hookimpl-decorator>`,
+plugins that are installed in the same python environment as napari can make
+themselves known to napari. ``napari`` will then scan plugin modules for *hook
+implementations* that will then be called at the appropriate time place during
+the execution of ``napari``.
+
+
+Step 1: Choose a hook specification to implement
+------------------------------------------------
+
+The functionality of plugins, as currently designed and implemented in
+``napari``, is rather specific in scope: They are *not* just independent code
+blocks with their own GUIs that show up next to the main napari window. Rather,
+plugin developers must decide which of the current *hook specifications*
+defined by napari that they would like to implement.
+
+For a complete list of *hook specifications* that developers can implement, see
+the :ref:`hook-specifications`.
+
+A single plugin package may implement more than one *hook specification*, but
+may not declare more the one *hook implementation* for any given specification.
+
+
+.. NOTE::
+   One of the primary ways that we will extend the functionality of napari over
+   time is by identifying new ideas for *hook specifications* that developers
+   can implement.  If you have a plugin idea that requires napari to create a
+   new hook specification, we'd love to hear about it!  Please think about what
+   the signature of your proposed hook specification would look like, and where
+   within the napari codebase you'd like your hook implementation to be called,
+   and `open a feature request
+   <https://github.com/napari/napari/issues/new?template=feature_request.md>`_
+   in the napari issue tracker with your proposal.
+
+Let's take the :func:`~napari.plugins.hook_specifications.napari_get_reader`
+hook (our primary "reader plugin" hook) as an example.  It is defined as:
+
+.. code-block:: python
+
+   LayerData = Union[Tuple[Any], Tuple[Any, Dict], Tuple[Any, Dict, str]]
+   ReaderFunction = Callable[[str], List[LayerData]]
+
+   @napari_hook_specification(firstresult=True)
+   def napari_get_reader(
+       path: Union[str, List[str]]
+   ) -> Optional[ReaderFunction]:
+       ...
+
+Note that it takes a ``str`` or a ``list`` of ``str`` and either returns
+``None`` or a function.  From the docstring of the hook specification, we see
```

Comment:
Can you include this docstring here? Or is it too much? Maybe you can link the function here also. (I see the link above the code block, but I would expect "docstring" to be clickable here. =)